File: erlang/seminars/philosophers/src/chop.erl
%% The two states of the chopstick.

available() ->
    receive 
	%% The first solution.
	{request, From} ->
	    From ! granted,
	    gone(na);


	%% Keeping track of requests.
	{request, Ref, From} ->
	    From ! {granted, Ref},
	    gone(Ref);


	quit ->
	    ok
    end.

gone(Ref) ->
    receive 
	%% The first solution
	return ->
	    available();

	%% Keeping track of request
	{return, Ref} ->
	    available();

	quit ->
	    ok
    end.

File: examples/phil/chop.ex
defmodule Chop do

  def start() do
    spawn_link(fn() -> available() end)
  end

  def request(chop) do
    send(chop, {:request, self()})
    receive do
      :granted ->
	:ok
    end
  end

  def return(chop) do
    send(chop, :return)
  end
  
  def available() do
    receive do
      {:request, from} ->
	send(from, :granted)
	gone()
      :quit ->
	:ok
    end
  end

  def gone() do
    receive do
      :return ->
	available()
      :quit ->
	:ok
    end
  end
  

end
